src/rvsdg/egglog_optimizer/loop_strength_red.egg

```diff
@@ -0,0 +1,76 @@
+;; in-loop
+;;
+;; ORIGINAL
+;; a = 0
+;; c = 3
+;; for  i = 0 to n:
+;;     a = i * c
+;;
+;; OPTIMIZED
+;; a = 0
+;; c = 3
+;; d = 0
+;; for i = 0 to n:
+;;     a += d
+;;     d += c
+
+(ruleset loop-strength-reduction)
+
+;; For a      theta, operand  is in the theta and either invariant or constant
+(relation inv (Body Operand))
+(rule (
+    (= theta (Theta pred inputs outputs))
+    (= (VecOperand-get outputs i) (Arg i)))
+    ((inv theta (Arg i))))
+(rule (
+    (= c (Node (PureOp (Const (IntT) (const) num))))
+    (= theta (Theta pred inputs outputs))
+    (Body-contains-Operand theta contains-i c)
+    )
+    ((inv theta c)))
+
+(rule 
+    (
+        ;; Find theta
+        (= old-theta (Theta pred (VO inputs) (VO outputs)))
+
+        ; Find invariant where input is same as output, or constant
+        (inv old-theta c)
+
+        ; Find loop variable (argument that gets incremented with an invariant)
+        (inv old-theta loop-incr)
+        (= (VecOperand-get (VO outputs) i) (Node (PureOp (badd (IntT) (Arg i) loop-incr))))
+
+        ; Find multiplication of loop variable and invariant
+        (= old-exp (Node (PureOp (bmul (IntT) c (Arg i)))))
+        (Body-contains-Operand old-theta contains-i old-exp)
+    )
+    (
+        ; Each time we need to update d by the product of the multiplied constant and the loop increment
+        (let addend (Node (PureOp (bmul (IntT) c loop-incr))))
+        ; n is index of our new, temporary variable d
+        (let n (vec-length inputs))
+
+        ; Initial value of d is i * c
+        (let d-init (Node (PureOp (bmul (IntT) c (VecOperand-get (VO inputs) i)))))
```

Comment:
Did you test this on an invariant argument instead of a constant? Doesn't it seem like you need to substitute here since the initial value is outside the loop?

Reply:
Yes, invariant argument does not work currently.